lib/onfleet/resource.rb

```diff
@@ -96,13 +99,85 @@ def self.delete(action = "", values = {})
     def self.request(uri, options = {})
       options = Nestful::Helpers.deep_merge(self.options, options)
 
-      response = Nestful::Request.new(
-        uri, options
-      ).execute
+      begin
+        response = Nestful::Request.new(
+          uri, options
+        ).execute
+      rescue JSON::ParserError
+        raise Errors::Error
+      rescue Nestful::ResponseError => e
+        handle_api_error(e)
+      end
 
       response
     end
 
+    def self.handle_api_error(nestful_error)
+      response = nestful_error.response
+
+      begin
+        error = JSON.parse(response.body)
+        error = error.with_indifferent_access if error
+        raise Errors::Error unless error && error.is_a?(Hash)
+      rescue JSON::ParserError, Errors::Error
+        raise general_error(response.code, response.body)
+      end
+
+      err_args = [error, response, nestful_error]
+
+      case response.code
+      when "400"
+        raise create_error Errors::InvalidContent, *err_args
+      when "401"
+        raise create_error Errors::InvalidCredentials, *err_args
+      when "404"
+        raise create_error Errors::ResourceNotFound, *err_args
+      when "409"
+        raise create_error Errors::PreconditionFailed, *err_args
+      when "412"
+        raise create_error Errors::PreconditionFailed, *err_args
+      when "429"
+        raise create_error Errors::TooManyRequests, *err_args
+      when "500"
+        raise create_error Errors::InternalError, *err_args
+      else
+        raise create_error Errors::OnfleetError, *err_args
+      end
+    end
+
+    def self.general_error(response_code, response_body)
+      Errors::Error.new(
+        message: "Invalid response object from API: #{response_code.inspect} " + "(HTTP response code was #{response_code})",
```

Comment:
Response code shows up twice, maybe shorten to just "Onfleet API failed with #{response_code}"?